### ISSUE TITLE
Plugin ABI Compatability

### ISSUE THREAD

**Issue by grigorescu** [zeek/zeek#1439]
With the work done in #266, we now have a stable and well-defined plugin API. However, ABI compatibility is lacking. If our API should be stable across Zeek major versions, I'd like to build a plugin for Zeek 4.0.0 and have it continue to work for 4.X.

Right now, we use the contents of the VERSION file to build an identifier: 
https://github.com/zeek/zeek/blob/85a0936ec6dc11406dd40546753d3afdaf276614/CMakeLists.txt#L93:L99

And then we do: https://github.com/zeek/zeek/blob/85a0936ec6dc11406dd40546753d3afdaf276614/zeek-config.h.in#L243

This means that any time the VERSION file changes, I have to rebuild my plugins. Following master, this happens pretty often (~daily).

This might also be a good time to rename the symbol from `bro` to `zeek`?

**Comment by jsiwek** [comment]
My understanding of C++ ABI stability is such that the design of Zeek's plugin API doesn't make the hypothetical effort of ABI stability any more easier/efficient: there's myriad API-stable changes to things in `zeek` namespace that wouldn't be ABI-stable, and also we wouldn't necessarily be free to make arbitrary changes to things in `detail` without considering impact it has on ABI-stability.

I've typically referenced [KDE's C++ ABI-stability guidelines](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B) when wondering about this and generally find many ABI-unsafe changes to be unintuitive/easy-to-forget and any mistakes made here are more likely to cause strange bugs of the undefined-runtime-behavior nature.

There may be tools to help automate ABI-compliance checks.  This is one, but I've no experience using it: https://lvc.github.io/abi-compliance-checker/.  Even using that, I think an effort to maintain ABI-stability with a C++-based API may be entering a world of pain where you frequently find: "that change is ABI-unsafe, plugins could break if we do it".  A C-based API would have significantly less rules and be easier to keep ABI-safe, but it would be an entirely new/re-written API and think it mostly comes out as:

* only use non-inlined, plain-C functions
* don't pass stack-allocated structs to functions, only heap-allocated (also note possible performance implications of this)
* don't re-order/remove struct fields
* really hide visibility of most symbols: go from "use things in `detail` at your possible expense of time-spent-maintaining" to "no really, you can only use these given APIs".

To summarize: I'd have little confidence in my ability to manually maintain ABI-stability of the current C++ API without automated tool-assistance.  Even then, it's an undertaking to carefully consider before committing to it since it comes with much higher maintenance costs and sacrifices overall development speed and a C-based API rather than a C++-based one would likely be more suitable/efficient in the end and something considered for an entire API rewrite again.

**Comment by grigorescu** [comment]
Thanks, Jon. Appreciate the pointer to the KDE guidelines, and I agree with your assessment.

I initially opened this issue specifically with `BRO_VERSION_FUNCTION` in mind. I then tried loosening that and quickly found that many other changes, specifically changes to `Plugin.h` will break the ABI.

Something else that I overlooked is that it's not realistic to expect compatibility for 4.X. Some functionality was deprecated in 4.0.0, and then removed in preparation of 4.1.0, in accordance with the published policy. So, when I tested between 4.0.0 and today's master, there were simply too many changes.

I'm curious what you think about the change in https://github.com/zeek/zeek/commit/35d6d78e2696a51b0422b18100ff2607c82a3164 . I modify `BRO_VERSION_FUNCTION` to only include Zeek's major version, and the plugin API's version. Perhaps Zeek's minor version should be part of this too. That's not a magic bullet to ABI compatibility, but it does at least remove changing the VERSION as being a breaking change. From then on, there are no guarantees, but realistically, master from today will mostly remain compatible with master from last week.

I could see that opening the door to being able to distribute pre-built binaries for plugins. Some plugins I've written are just a pain to build -- you have to install other dependencies, etc. With taking a "best effort" at compatibility, I could see a nightly CI job that builds against master, and only publishes a new version if the symbols have changed.

**Comment by rsmmr** [comment]
I can see loosening this up a bit, but would rather not go too far with that. For one, I'd leave VERSION_MIINOR in; it's more likely than not that ABI stuff changes between minor versions. But I can see weakening it up for the patch level. Maybe for master, we don't need to include the patch level at all. For release version, I'm wondering if it would be feasible to make  the decision manually: we'd have a "release ABI version" that we increase if making an ABI-breaking change when the patch level increases. I imagine that for a patch-level release that wouldn't be very often; and I think Jon's (valid) concern of not even being able to tell if something breaks ABI, would be less of an issue with changes at the level of security updates.

**Comment by grigorescu** [comment]
> I can see loosening this up a bit, but would rather not go too far with that. For one, I'd leave VERSION_MIINOR in; it's more likely than not that ABI stuff changes between minor versions. But I can see weakening it up for the patch level. Maybe for master, we don't need to include the patch level at all.

Ok, I can add minor version in.

Right now, we also distinguish based on whether or not it's a debug build. I don't really see a need to keep that in?

**Comment by 0xxon** [comment]
I actually think we have some (minor) ABI changes from debug/non-debug builds. I admittedly don't recall the details, but I think there is at least one place where a function does not exist when we are not in a debug build.

**Comment by rsmmr** [comment]
Yeah, I don't think plugins are generally compatible between debug/non-debug builds. The debug logging code compiles conditionally.

**Comment by 0xxon** [comment]
Also - just to repeat Robin a bit - I don't think we should loosen this up too much. We actually tightened this for a reason - we, especially at the moment, introduce quite a bit of ABI incompatibilities. And that can lead to hard to diagnose problems.

Would it, instead, perhaps make more sense to have a way to more easily get plugins recompiled, or even have them recompiled automatically?

**Comment by grigorescu** [comment]
> Would it, instead, perhaps make more sense to have a way to more easily get plugins recompiled, or even have them recompiled automatically?

Yes and no. As soon as #1441 gets merged, I can include plugins in our current deployment workflow. However, this wouldn't help plugin developers that want to offer pre-built versions of their plugins. This would be most useful for those plugins which have extra dependencies that need to be installed in order to build.

If we can loosen this, then I intend to add some GitHub Actions to the Zeek Exporter plugin, for instance, to offer an easier installation mechanism.

**Comment by 0xxon** [comment]
Hm. I am repeating myself here - but I am really not sure about loosening this much. Jon mentioned it a bit above - it is really easy to make ABI incompatible changes - and we probably have quite a few of them. If we just loosen this, while continuing to make these changes, it is quite likely that this will lead to crashes with really weird backtraces for people running such plugins. Which also is not great due to a whole bunch of reasons...

I basically think we should make it as easy as possible for plugins to be built from source. And if someone wants to distribute pre-built versions of their plugins - they should probably just distribute them for one specific release.

**Comment by grigorescu** [comment]
I guess I'm confused about the incompatibility that you're referring to. What I'm proposing is that we loosen this intentional ABI incompatibility. If that symbol is no longer an issue, the loader still detects and errors out on any other ABI incompatibility, doesn't it? It's not as if we're saying "ABI compatibility no longer matters," we're just trusting the loader to do its job and we're not arbitrarily breaking compatibility when it might otherwise be compatible.

Today, if I build a plugin against master from last week, and then try to run it against master from today, it will fail to load with an error message along the lines of:

`undefined symbol: bro_version_4_1_0_dev_323_plugin_7`

If we loosen this to `bro_version_4_1_plugin_7`, with the same scenario as above, we have two options. Either there were no ABI incompatibilities introduced between master from last week and master from this week, in which case the plugin works as intended. Or, some incompatibilities were introduced, in which case I get something like:

`undefined symbol: _ZN4zeek6plugin6Plugin16HookCallFunctionEPKNS_4FuncEPNS_6detail5FrameEPNS_4ListIPNS_3ValELNS_9ListOrderE0EEE`

It seems to me that the `bro_version` symbol serves two purposes. First, it's a less "scary" error message for users to see. I'm not convinced about the value of this, because we don't exactly document what "undefined symbol: bro_version" means anywhere, nor is it easy for a user to figure out which symbol the plugin is looking for versus which plugin Zeek has.

The second use-case is the more insidious one. It's possible that ABI compatibility is maintained between two versions, but something changes at a deeper level that would cause issues. I can't really think of examples of this. Even all the recent IntrusivePtr surgery, and other significant behind-the-scenes changes have led to new function names, or BIFs that were changed in coordination with the deprecation policy.

> I basically think we should make it as easy as possible for plugins to be built from source. And if someone wants to distribute pre-built versions of their plugins - they should probably just distribute them for one specific release.

I don't disagree, but my work is specifically in response to Robin's comments about how we need more people running master, and that during the 4.0 release cycle, and specifically when we had release candidates, it was very hard to get good tests with live traffic. So, I'm trying to focus specifically on making it easier for people to run master, and intentionally breaking it with every version update just adds more hurdles.

**Comment by 0xxon** [comment]
> If that symbol is no longer an issue, the loader still detects and errors out on any other ABI incompatibility, doesn't it? 

The problem is that - no, a lot of incompatibilities cannot be detected by the loader. They instead will lead to (sometimes quite hard to debug and sometimes random) crashes or weird behavior.

You mention this a bit further down - the reason that we introduced this error message and this intentional incompatibility is that exactly these kinds of problems (loader was happy loading everything, but crashes happened) did indeed happen in practice.

I admittedly don't quite remember what caused them - but it is probably smaller changes (think - a function slightly changing the arguments that it takes, without changing the function name - which is a quite common change).

**Comment by grigorescu** [comment]
Ok, that makes sense. I was getting hung up on "ABI incompatible changes" only breaking the ABI itself, hence my confusion. On the same page now.

To bring this back on track, a few concrete questions:

1. Are these breaking changes something that could be detected via btests? Or is that impractical because one would need to essentially test every function?
2. Is there a mechanism that we can provide to plugin developers that would allow them to override this check? I could envision a workflow where a plugin developer has a CI/CD pipeline that loads yesterday's version of their plugin binary into Zeek master, runs their test suite, and if their tests pass, they'd be satisfied that no updates are needed. Certainly, this would require a comprehensive test suite.
3. Should we deprecate the `bro_version` symbol and introduce a `zeek_version` symbol?

**Comment by jsiwek** [comment]
> 1. Are these breaking changes something that could be detected via btests? Or is that impractical because one would need to essentially test every function?

Via general automation/CI: maybe crudely or partially at the moment, but I'd still have doubts.  For example, there's tools for checking [ABI compliance](https://lvc.github.io/abi-compliance-checker/), but that one does not mention being able to detect a subtle incompatibility like making a semantic change to an inlined function and not sure it ever could help there.  Maybe those types of changes are rare, but seems like true automated detection of it does require 100% unit test coverage.

> 2\. Is there a mechanism that we can provide to plugin developers that would allow them to override this check?

Probably, but don't think I'd suggest using such a thing.

**Comment by grigorescu** [comment]
Ok. I'll close this, then.

**Comment by rsmmr** [comment]
Let me reopen to ask folks about my proposal above because I think that could be a helpful compromise.

We could define BRO_VERSION_FUNCTION like this:

1. For release versions: `MAJOR_MINOR_ABI` where `ABI` is a simple counter that we increase manually with bug fix releases (only) when we have to.
2. For master: `MAJOR_MINOR_dev` (no patchlevel).

I'm guessing that this would take care of most unnecessary plugin recompiles without causing the users additional trouble (unless they are on master, in which case they are on their own anyways).

**Comment by jsiwek** [comment]
> * For release versions: `MAJOR_MINOR_ABI` where `ABI` is a simple counter that we increase manually with bug fix releases (only) when we have to.

It's extra release/maintenance overhead and failure-point for incorrect manual decisions about whether to bump ABI version.  Otherwise, yes, that's a way to reduce unnecessary plugin rebuilds if people find them problematic/inconvenient.

> * For master: `MAJOR_MINOR_dev` (no patchlevel).

That may reduce some unnecessary plugin rebuilds, but if it doesn't help catch the times when a rebuild is needed for intended/correct behavior and resulting misbehaviors aren't always trivial to detect, people that want predictable software behavior will rebuild for every master-branch change via automation or manual obsessiveness anyway?

If it's a personal preference tradeoff between potential convenience and predictability, I'd say at least don't make the default behavior be the less predictable one, maybe have an opt-in `./configure` flag for the less-strict ABI checking method.

**Comment by 0xxon** [comment]
I agree with @jsiwek here - I am worried about the failure cases here and would prefer this not to be the default behavior.

**Comment by rsmmr** [comment]
Sounds like we don't have enough support for changes here, re-closing.
